README.md
mysql_community_release Cookbook
================================
This cookbook is extend the mysql cookbook with the "MySQL Community Server"

Requirements
------------

#### packages
- `mysql` - mysql_community_release needs mysql to use.

Attributes
----------

#### mysql_community_release::default

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['mysql_community_release']['rpm_path']</tt></td>
    <td>String</td>
    <td>RPM version of mysql community server</td>
    <td><tt>mysql-community-release-el6-5.noarch.rpm</tt></td>
  </tr>
</table>

Usage
-----

#### mysql_community_release::default
Just include `mysql_community_release` in your node's `run_list`:

```json
{
  "mysql": {
    "version": "5.6",
    "server_root_password": "supersecretpassword",
    "server_repl_password": "supersecretpassword"
  }
  "run_list": [
    "recipe[mysql_community_release]"
  ]
}
```

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: Yusuke Tanaka (Akatsuki, Inc.)

under MIT License


CHANGELOG.md
mysql_community_release CHANGELOG
=================================

This file is used to list changes made in each version of the mysql_community_release cookbook.

0.1.0
-----
- [Yusuke Tanaka] - Initial release of mysql_community_release

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


libraries/helpers.rb
module Opscode
  module Mysql
    class PlatformInfo
      def self.mysql_info
        @mysql_info ||= {
          'rhel' => {
            'default_data_dir' => '/var/lib/mysql',
            '5' => {
              'default_version' => '5.0',
              '5.0' => 'mysql-server',
              '5.1' => 'mysql51-mysql-server',
              '5.5' => 'mysql55-mysql-server'
            },
            '6' => {
              'default_version' => '5.1',
              '5.1' => 'mysql-server'
            },
            '7' => {
              'default_version' => '5.5',
              '5.1' => 'mysql51-server',
              '5.5' => 'mysql55-server'
            },
            '2013.09' => {
              'default_version' => '5.6',
              '5.1' => 'mysql-server',
              '5.6' => 'mysql-community-server'
            },
            '2014.03' => {
              'default_version' => '5.6',
              '5.1' => 'mysql51-server',
              '5.5' => 'mysql55-server',
              '5.6' => 'mysql-community-server'
            }
          },
          'fedora' => {
            'default_data_dir' => '/var/lib/mysql',
            '19' => {
              'default_version' => '5.5',
              '5.5' => 'community-mysql-server'
            },
            '20' => {
              'default_version' => '5.5',
              '5.5' => 'community-mysql-server'
            }
          },
          'debian' => {
            'default_data_dir' => '/var/lib/mysql',
            '7' => {
              'default_version' => '5.5',
              '5.5' => 'mysql-server-5.5'
            },
            '10.04' => {
              'default_version' => '5.1',
              '5.1' => 'mysql-server-5.1'
            },
            '12.04' => {
              'default_version' => '5.5',
              '5.5' => 'mysql-server-5.5'
            },
            '13.10' => {
              'default_version' => '5.5',
              '5.5' => 'mysql-server-5.5'
            }
          },
          'smartos' => {
            'default_data_dir' => '/opt/local/lib/mysql',
            # Do this or now, until Ohai correctly detects a
            # smartmachine vs global zone (base64 13.4.0) from /etc/product
            '5.11' => {
              'default_version' => '5.5',
              '5.5' => 'mysql-server',
              '5.6' => 'mysql-server'
            }
          },
          'omnios' => {
            'default_data_dir' => '/var/lib/mysql',
            '151006' => {
              'default_version' => '5.5',
              '5.5' => 'database/mysql-55',
              '5.6' => 'database/mysql-56'
            }
          }
        }
      end
    end
  end
end


libraries/provider_mysql_service_rhel.rb
require 'chef/provider/lwrp_base'

class Chef
  class Provider
    class MysqlService
      class Rhel < Chef::Provider::MysqlService
        use_inline_resources if defined?(use_inline_resources)

        def whyrun_supported?
          true
        end

        action :create do
          case node['platform_version'].to_i.to_s
          when '2013'
            case new_resource.version
            when '5.1'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-server'
              service_name = 'mysqld'
            when '5.6'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-community-server'
              service_name = 'mysqld'
            end
          when '2014'
            case new_resource.version
            when '5.1'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-server'
              service_name = 'mysqld'
            when '5.5'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-server'
              service_name = 'mysqld'
            when '5.6'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-community-server'
              service_name = 'mysqld'
            end
          when '6'
            case new_resource.version
            when '5.1'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-server'
              service_name = 'mysqld'
            when '5.6'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-community-server'
              service_name = 'mysqld'
            end
          when '5'
            case new_resource.version
            when '5.0'
              base_dir = ''
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/usr'
              lc_messages_dir = nil
              run_dir = '/var/run/mysqld'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql-server'
              service_name = 'mysqld'
            when '5.1'
              base_dir = '/opt/rh/mysql51/root'
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/opt/rh/mysql51/root/usr'
              lc_messages_dir = nil
              run_dir = '/opt/rh/mysql51/root/var/run/mysqld/'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql51-mysql-server'
              service_name = 'mysql51-mysqld'
            when '5.5'
              base_dir = '/opt/rh/mysql55/root'
              include_dir = "#{base_dir}/etc/mysql/conf.d"
              prefix_dir = '/opt/rh/mysql55/root/usr'
              lc_messages_dir = nil
              run_dir = '/opt/rh/mysql55/root/var/run/mysqld/'
              pid_file = '/var/run/mysql/mysql.pid'
              socket_file = '/var/lib/mysql/mysql.sock'
              package_name = 'mysql55-mysql-server'
              service_name = 'mysql55-mysqld'
            end
          end

          converge_by 'rhel pattern' do
            package package_name do
              action :install
            end

            directory include_dir do
              owner 'mysql'
              group 'mysql'
              mode '0750'
              recursive true
              action :create
            end

            directory run_dir do
              owner 'mysql'
              group 'mysql'
              mode '0755'
              recursive true
              action :create
            end

            directory new_resource.data_dir do
              owner 'mysql'
              group 'mysql'
              mode '0750'
              recursive true
              action :create
            end

            service service_name do
              supports :restart => true
              action [:start, :enable]
            end

            execute 'wait for mysql' do
              command "until [ -S #{socket_file} ] ; do sleep 1 ; done"
              timeout 10
              action :run
            end

            template '/etc/mysql_grants.sql' do
              cookbook 'mysql'
              source 'grants/grants.sql.erb'
              owner  'root'
              group  'root'
              mode   '0600'
              action :create
              notifies :run, 'execute[install-grants]'
            end

            if node['mysql']['server_root_password'].empty?
              pass_string = ''
            else
              pass_string = "-p#{node['mysql']['server_root_password']}"
            end

            execute 'install-grants' do
              cmd = "#{prefix_dir}/bin/mysql"
              cmd << ' -u root '
              cmd << "#{pass_string} < /etc/mysql_grants.sql"
              command cmd
              action :nothing
            end

            template "#{base_dir}/etc/my.cnf" do
              if new_resource.template_source.nil?
                source "#{new_resource.version}/my.cnf.erb"
                cookbook 'mysql'
              else
                source new_resource.template_source
              end
              owner 'mysql'
              group 'mysql'
              mode '0600'
              variables(
                :base_dir => base_dir,
                :data_dir => new_resource.data_dir,
                :include_dir => include_dir,
                :lc_messages_dir => lc_messages_dir,
                :pid_file => pid_file,
                :port => new_resource.port,
                :socket_file => socket_file
                )
              action :create
              notifies :run, 'bash[move mysql data to datadir]'
              notifies :restart, "service[#{service_name}]"
            end

            bash 'move mysql data to datadir' do
              user 'root'
              code <<-EOH
              service #{service_name} stop \
              && for i in `ls #{base_dir}/var/lib/mysql | grep -v mysql.sock` ; do mv #{base_dir}/var/lib/mysql/$i #{new_resource.data_dir} ; done
              EOH
              action :nothing
              only_if "[ '#{base_dir}/var/lib/mysql' != #{new_resource.data_dir} ]"
              only_if "[ `stat -c %h #{new_resource.data_dir}` -eq 2 ]"
              not_if "[ `stat -c %h #{base_dir}/var/lib/mysql/` -eq 2 ]"
            end

            execute 'assign-root-password' do
              cmd = "#{prefix_dir}/bin/mysqladmin"
              cmd << ' -u root password '
              cmd << node['mysql']['server_root_password']
              command cmd
              action :run
              only_if "#{prefix_dir}/bin/mysql -u root -e 'show databases;'"
            end
          end
        end

        action :restart do
          # FIXME: Find a way to DRY this
          case node['platform_version'].to_i.to_s
          when '2013'
            case new_resource.version
            when '5.1', '5.6'
              service_name = 'mysqld'
            end
          when '2014'
            case new_resource.version
            when '5.1', '5.5', '5.6'
              service_name = 'mysqld'
            end
          when '6'
            case new_resource.version
            when '5.1', '5.6'
              service_name = 'mysqld'
            end
          when '5'
            case new_resource.version
            when '5.0'
              service_name = 'mysqld'
            when '5.1'
              service_name = 'mysql51-mysqld'
            when '5.5'
              service_name = 'mysql55-mysqld'
            end
          end

          converge_by 'rhel pattern' do
            service service_name do
              supports :restart => true
              action :restart
            end
          end
        end
      end
    end
  end
end

Chef::Platform.set :platform => :amazon, :resource => :mysql_service, :provider => Chef::Provider::MysqlService::Rhel
Chef::Platform.set :platform => :redhat, :resource => :mysql_service, :provider => Chef::Provider::MysqlService::Rhel
Chef::Platform.set :platform => :centos, :resource => :mysql_service, :provider => Chef::Provider::MysqlService::Rhel
Chef::Platform.set :platform => :oracle, :resource => :mysql_service, :provider => Chef::Provider::MysqlService::Rhel
Chef::Platform.set :platform => :scientific, :resource => :mysql_service, :provider => Chef::Provider::MysqlService::Rhel
